### PR TITLE
[sf2.7+] Add choice_loader options for multi choice field

### PR DESCRIPTION
**Pull request by Mangetsu**
Add choice_loader fix for symfony 2.7+ to avoid exception when using a multichoice form widget

### CODE REVIEW

**Comment by Mangetsu** [comment]
@greg0ire I removed and made a new branch with clean code now, let me know if it's ok now.

**Comment by greg0ire** [comment]
It looks way better :)

**Comment by greg0ire** [comment]
About the build : not sure it is your fault, but `make cs` should make it pass. If it does not, please show us the output of `make cs`.

**Comment by Mangetsu** [comment]
I guess i didn't add unit tests or something (please be patient with me it's my first time contributing :cold_sweat: 

`> Starting Tests

Loaded config from "/project/.php_cs"
..........................................................................................................................................................................II......................................................................................................................................F................F...........
Legend: ?-unknown, I-invalid file syntax, file ignored, .-no changes, F-fixed, E-error
   1) Form/Type/ModelType.php (trailing_spaces)
   2) Form/DataTransformer/ModelsToArrayTransformer.php (single_quote, braces)
Checked all files in 136.660 seconds, 23.000 MB memory used
ERR: Error while validating convention standards
Build step 'Execute shell' marked build as failure
Recording test results
ERROR: Publisher 'Publish JUnit test result report' failed: No test report files were found. Configuration error?
[xUnit] [INFO] - Starting to record.
[xUnit] [INFO] - Processing JUnit
[xUnit] [INFO] - [JUnit] - No test report file(s) were found with the pattern 'build/junit.xml' relative to '/var/jenkins_home/jobs/sonata-ci-bundle-admin/workspace' for the testing framework 'JUnit'.  Did you enter a pattern relative to the correct directory?  Did you generate the result report(s) for 'JUnit'?
[xUnit] [ERROR] - No test reports found for the metric 'JUnit' with the resolved pattern 'build/junit.xml'. Configuration error?.
[xUnit] [INFO] - Failing BUILD.
[xUnit] [INFO] - There are errors when processing test results.
[xUnit] [INFO] - Skipping tests recording.
[xUnit] [INFO] - Stop build.
Notifying upstream projects of job completion
Setting status of 3effe4560316696fd4d25a81e758202af32f86b0 to FAILURE with url http://91.121.3.21/job/sonata-ci-bundle-admin/896/ and message: ' No test results found.'
Using context: SonataCI - Quick Check
Finished: FAILURE`

**Comment by greg0ire** [comment]
Can you run `make cs` locally ? Some files should change (git status to confirm that). The message you pasted above is about cs, not unit tests.

> please be patient with me it's my first time contributing

No problem, you can do it, you're very close!

**Comment by Mangetsu** [comment]
You'll surely laugh at me, because my dev env is limited to an instance of wamp + git CLI + PHPStorm as IDE, even unit tests are not part of my dev process :confused:

**Comment by greg0ire** [comment]
Oh I get it now. Have you considered switching to linux? Or using docker maybe? It could be great to avoid bugs that happen only in production. Anyway, the git CLI is based on msysgit if I'm not wrong, doesn't `make` work in it? I'm going to fetch your changes, run `make cs` and show you the diff.

**Comment by greg0ire** [comment]
> You'll surely laugh at me

I started on Windows too BTW, so no shame. Just make sure you try linux, it is really focused on development, and using it can make you more familiar with what you use in production (I guess it is not Windows). In the meantime, we are interesting in making this work for you under Windows : there are probably other contributors that use Windows too, and we would like them to be able to make their contributions just as easily as us.

**Comment by Mangetsu** [comment]
@greg0ire  thank you sir :smile:

**Comment by greg0ire** [comment]
Here is the promised diff : 

```patch
diff --git a/Form/DataTransformer/ModelsToArrayTransformer.php b/Form/DataTransformer/ModelsToArrayTransformer.php
index a47798f..bc85cc1 100644
--- a/Form/DataTransformer/ModelsToArrayTransformer.php
+++ b/Form/DataTransformer/ModelsToArrayTransformer.php
@@ -53,9 +53,8 @@ public function __construct($choiceList, ModelManagerInterface $modelManager, $c
         if (!$choiceList instanceof ModelChoiceList
             && !$choiceList instanceof ModelChoiceLoader
             && !$choiceList instanceof LazyChoiceList) {
-            throw new RuntimeException("First param passed to ModelsToArrayTransformer should be instance of 
-                ModelChoiceLoader or ModelChoiceList or LazyChoiceList");
-
+            throw new RuntimeException('First param passed to ModelsToArrayTransformer should be instance of 
+                ModelChoiceLoader or ModelChoiceList or LazyChoiceList');
         }
 
         $this->choiceList   = $choiceList;
diff --git a/Form/Type/ModelType.php b/Form/Type/ModelType.php
index d277d89..c6ab0bb 100644
--- a/Form/Type/ModelType.php
+++ b/Form/Type/ModelType.php
@@ -52,8 +52,8 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         if ($options['multiple']) {
             if (array_key_exists('choice_loader', $options) && $options['choice_loader'] !== null) { // SF2.7+
                 $builder->addViewTransformer(new ModelsToArrayTransformer(
-                    $options['choice_loader'], 
-                    $options['model_manager'], 
+                    $options['choice_loader'],
+                    $options['model_manager'],
                     $options['class']), true);
             } else {
                 $builder->addViewTransformer(new LegacyModelsToArrayTransformer($options['choice_list']), true);
```

Can you : 

- rework your commit (try to have still 2 commits in the end, using git rebase --interactive origin/master)
- tell us what happen when you run `make test` in your git-bash prompt, for science ?

File: Form/DataTransformer/ModelsToArrayTransformer.php [hunk 1]
```diff
@@ -50,8 +50,12 @@ class ModelsToArrayTransformer implements DataTransformerInterface
      */
     public function __construct($choiceList, ModelManagerInterface $modelManager, $class)
     {
-        if (!$choiceList instanceof ModelChoiceList && !$choiceList instanceof LazyChoiceList) {
-            throw new RuntimeException('First param passed to ModelsToArrayTransformer should be instance of ModelChoiceList or LazyChoiceList');
+        if (!$choiceList instanceof ModelChoiceList
+            && !$choiceList instanceof ModelChoiceLoader
+            && !$choiceList instanceof LazyChoiceList) {
+            throw new RuntimeException("First param passed to ModelsToArrayTransformer should be instance of
```
Comment (greg0ire) [inline Form/DataTransformer/ModelsToArrayTransformer.php]: You need to use single quotes here

File: Form/Type/ModelType.php [hunk 2]
```diff
@@ -51,7 +51,10 @@ public function buildForm(FormBuilderInterface $builder, array $options)
     {
         if ($options['multiple']) {
             if (array_key_exists('choice_loader', $options) && $options['choice_loader'] !== null) { // SF2.7+
-                $builder->addViewTransformer(new ModelsToArrayTransformer($options['choice_list'], $options['model_manager'], $options['class']), true);
+                $builder->addViewTransformer(new ModelsToArrayTransformer(
+                    $options['choice_loader'], 
+                    $options['model_manager'],
```
Comment (greg0ire) [inline Form/Type/ModelType.php]: There are trailing spaces at then end of lines 54 and 55. I think you can configure phpstorm to remove them on save.

**Pull request closed without merging**